Refactor Button component.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #10357 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/index.js:
```diff
@@ -97,25 +92,24 @@ const Button = forwardRef(
+			? title || customizedTooltip || ariaLabel
+			: null;
+
+		const buttonComponent = (
```
Since this variable is not reused anymore, we should get rid of it and render the component inline. What do you think?

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/MaybeTooltip.js:
```diff
@@ -0,0 +1,49 @@
+ */
+import Tooltip from '../Tooltip';
+
+const MaybeTooltip = ( {
```
Our convention is to define components using the `function` keyword. Let's do that here.

Example:

```js
export default function MaybeTooltip
```

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/SemanticButton.js:
```diff
@@ -0,0 +1,80 @@
+ */
+import { forwardRef } from '@wordpress/element';
+
+const SemanticButton = forwardRef(
```
Let's define prop types for this component, or somehow reuse from the parent component. While we're at it, it would be nice to update any outdated prop types.

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/MaybeTooltip.test.js:
```diff
@@ -0,0 +1,111 @@
+ * limitations under the License.
+ */
+
+/**
+ * MaybeTooltip tests.
+ *
+ * Site Kit by Google, Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Looks like this was duplicated.

```suggestion
```

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/MaybeTooltip.test.js:
Solid tests here, thank you!

A very nitpicky observation: We seem to be capitalising "Tooltip" in the test case names, which I think we shouldn't. Please let me know what you think, thanks!

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/SemanticButton.test.js:
```diff
@@ -0,0 +1,120 @@
+/**
+ * SemanticButton tests.
+ *
+ * Site Kit by Google, Copyright 2021 Google LLC
```
```suggestion
 * Site Kit by Google, Copyright 2025 Google LLC
```

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/SemanticButton.test.js:
```diff
@@ -0,0 +1,120 @@
+		).not.toBeInTheDocument();
+	} );
+
+	it( 'should apply danger class when danger prop is true', () => {
+		const { container } = render(
+			<SemanticButton danger>Test</SemanticButton>
+		);
+		expect( container.firstChild ).toHaveClass( 'mdc-button--danger' );
+	} );
+
+	it( 'should apply inverse class when inverse prop is true', () => {
+		const { container } = render(
+			<SemanticButton inverse>Test</SemanticButton>
+		);
+		expect( container.firstChild ).toHaveClass( 'mdc-button--inverse' );
+	} );
+
+	it( 'should apply tertiary class when tertiary prop is true', () => {
+		const { container } = render(
+			<SemanticButton tertiary>Test</SemanticButton>
+		);
+		expect( container.firstChild ).toHaveClass( 'mdc-button--tertiary' );
+	} );
+
+	it( 'should apply callout classes when callout props are set', () => {
+		const { container } = render(
+			<SemanticButton callout calloutStyle="warning">
+				Test
+			</SemanticButton>
+		);
+		expect( container.firstChild ).toHaveClass( 'mdc-button--callout' );
+		expect( container.firstChild ).toHaveClass(
+			'mdc-button--callout-warning'
+		);
+	} );
```
Not very important, but this could be a nice spot to use `it.each`.

nfmohit reviewed (changes requested):
Brilliant work here, thank you @ankitrox!

I've left a few minor comments for your consideration. Please let me if you have any questions or concerns, thank you!

ankitrox commented:
Thank you @nfmohit for reviewing the PR and adding your valuable suggestions.

I've addressed the feedback and sending it your way to review those.

Thanks again!

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/MaybeTooltip.js:
```diff
@@ -0,0 +1,72 @@
+	tooltipEnterDelayInMS: PropTypes.number,
+};
+
+MaybeTooltip.defaultProps = {
+	disabled: false,
+	tooltip: false,
+	tooltipTitle: null,
+	hasIconOnly: false,
+	tooltipEnterDelayInMS: 100,
+};
```
Let's use default parameters in the exported function, as `defaultProps` on functional components will eventually be deprecated (see [RFC](https://github.com/reactjs/rfcs/pull/107)).

With that said, as we're already refactoring, we should apply this refactor to the parent component too, if capacity permits, thank you!

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/SemanticButton.js:
```diff
@@ -0,0 +1,103 @@
+	}
+);
+
+SemanticButton.displayName = 'SemanticButton';
```
Sorry for not flagging this earlier, but I don't think the `displayName` property is necessary here, as it should ideally only be used when the component should be labelled differently than the exported function.

This should also be removed from the parent component, IMO.

nfmohit commented on assets/js/googlesitekit/components-gm2/Button/SemanticButton.js:
```diff
@@ -0,0 +1,103 @@
+
+SemanticButton.displayName = 'SemanticButton';
+
+SemanticButton.propTypes = {
+	children: PropTypes.node,
+	href: PropTypes.string,
+	text: PropTypes.bool,
+	className: PropTypes.string,
+	danger: PropTypes.bool,
+	disabled: PropTypes.bool,
+	target: PropTypes.string,
+	icon: PropTypes.element,
+	trailingIcon: PropTypes.element,
+	'aria-label': PropTypes.string,
+	inverse: PropTypes.bool,
+	tertiary: PropTypes.bool,
+	callout: PropTypes.bool,
+	calloutStyle: PropTypes.oneOf( [ 'primary', 'warning', 'error' ] ),
+	title: PropTypes.string,
+	customizedTooltip: PropTypes.element,
+	tooltip: PropTypes.bool,
+	hideTooltipTitle: PropTypes.bool,
+	tooltipEnterDelayInMS: PropTypes.number,
+};
```
I don't think `icon`, `trailingIcon`, `title`, `customizedTooltip`, `tooltip`, `hideTooltipTitle`, and `tooltipEnterDelayInMS` are valid props for this component.

```suggestion
SemanticButton.propTypes = {
	children: PropTypes.node,
	href: PropTypes.string,
	text: PropTypes.bool,
	className: PropTypes.string,
	danger: PropTypes.bool,
	disabled: PropTypes.bool,
	target: PropTypes.string,
	'aria-label': PropTypes.string,
	inverse: PropTypes.bool,
	tertiary: PropTypes.bool,
	callout: PropTypes.bool,
	calloutStyle: PropTypes.oneOf( [ 'primary', 'warning', 'error' ] ),
};
```

nfmohit reviewed (changes requested):
Thank you for addressing my feedback, @ankitrox!

I've left a few last-minute comments for your consideration. Please let me know if you have any questions or concerns, thank you!

nfmohit reviewed (approved):
LGTM, nice work, @ankitrox! ✅
